### DISCUSSION TITLE
Decompile to JavaScript

### DISCUSSION THREAD

**Discussion by mrpmorris** [icsharpcode/ILSpy#3214] [Ideas] [answered]
I really love the output of the decompiler from IL to C#, the output is very natural looking, well done!

This might be a weird thought, but I am wondering it it would be possible to create a decompiler to JavaScript?

I'm thining of an equivalent of a C# equivalent of Fable for F#. It would be nice to be able to write web apps in C# without having to use Web Assembly.

**Comment by christophwille** [comment]
Decompile to JS might not be the best path - going from valid C# (yes, that is what we focus on) to JS should be the better option.

**Reply by mrpmorris** [reply]
That's what I mean. 

Like Fable is for F#.

**Reply by siegfriedpammer** [reply, marked as answer]
Yes, you probably want something like https://github.com/SharpKit/SharpKit (a C#-to-JavaScript compiler) or https://github.com/sq/JSIL (an MSIL-to-JavaScript compiler). This is out-of-scope of the ILSpy project, because we focus on displaying .NET IL byte-code and assembly metadata as well as trying to reconstruct the original C# code from IL by detecting and reversing patterns commonly used by C# compilers. Note that both of the previously mentioned projects seem to be abandoned, but the second one even uses an old version (pre-v3.0?) of the ILSpy decompiler engine to get a higher-level view of the IL byte-code.

My impression is that .NET-to-JavaScript is somewhat dead since the advent of WebAssembly, which is now supported as a first-class target by the Microsoft tool chain. Also it's a lot easier and faster to translate one stack-based environment to another one, than going from a more high-level language like C# to another high-level language like JavaScript with vastly different semantics. Stack-based systems usually have very similar instruction sets with simpler semantics.

**Reply by mrpmorris** [reply]
Thank you for the information!
